src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs

```diff
@@ -1222,4 +1223,1555 @@ struct S3
+        CompileAndVerify(comp).VerifyDiagnostics();
+    }
+
+    [Fact]
+    public void EnforcedRequiredMembers_OverrideRetargeted_AllSet()
+    {
+        var retargetedCode = @"public class C {}";
+        var originalC = CreateCompilation(new AssemblyIdentity("Ret", new Version(1, 0, 0, 0), isRetargetable: true), retargetedCode, TargetFrameworkUtil.StandardReferences);
+
+        var @base = @"
+public class Base
+{
+    public required virtual C Prop1 { get; set; }
+}
+";
+
+        var baseComp = CreateCompilationWithRequiredMembers(@base, new[] { originalC.ToMetadataReference() }, targetFramework: TargetFramework.Standard);
+        var retargetedC = CreateCompilation(new AssemblyIdentity("Ret", new Version(2, 0, 0, 0), isRetargetable: true), retargetedCode, TargetFrameworkUtil.StandardReferences);
+
+        var code = @"
+_ = new Derived() { Prop1 = new() };
+
+public class Derived : Base
+{
+    public override required C Prop1 { get; set; }
+}
+";
+
+        var comp = CreateCompilation(code, new[] { baseComp.ToMetadataReference(), retargetedC.ToMetadataReference() });
+        comp.VerifyDiagnostics();
+
+        Assert.IsType<RetargetingNamedTypeSymbol>(comp.GetTypeByMetadataName("Base"));
+    }
+
+    [Fact]
+    public void EnforcedRequiredMembers_OverrideRetargeted_NoneSet()
+    {
+        var retargetedCode = @"public class C {}";
+        var originalC = CreateCompilation(new AssemblyIdentity("Ret", new Version(1, 0, 0, 0), isRetargetable: true), retargetedCode, TargetFrameworkUtil.StandardReferences);
+
+        var @base = @"
+public class Base
+{
+    public required virtual C Prop1 { get; set; }
+}
+";
+
+        var baseComp = CreateCompilationWithRequiredMembers(@base, new[] { originalC.ToMetadataReference() }, targetFramework: TargetFramework.Standard);
+        var retargetedC = CreateCompilation(new AssemblyIdentity("Ret", new Version(2, 0, 0, 0), isRetargetable: true), retargetedCode, TargetFrameworkUtil.StandardReferences);
+
+        var code = @"
+_ = new Derived();
+
+public class Derived : Base
+{
+    public override required C Prop1 { get; set; }
+}
+";
+
+        var comp = CreateCompilation(code, new[] { baseComp.ToMetadataReference(), retargetedC.ToMetadataReference() });
+        comp.VerifyDiagnostics(
+            // (2,9): error CS9506: Required member 'Derived.Prop1' must be set in the object initializer or attribute constructor.
+            // _ = new Derived();
+            Diagnostic(ErrorCode.ERR_RequiredMemberMustBeSet, "Derived").WithArguments("Derived.Prop1").WithLocation(2, 9)
+        );
+
+        Assert.IsType<RetargetingNamedTypeSymbol>(comp.GetTypeByMetadataName("Base"));
+    }
+
+    [Fact]
+    public void EnforcedRequiredMembers_ShadowedInSource_01()
+    {
+        var vb = @"
+Imports System.Runtime.CompilerServices
+<RequiredMember>
+Public Class Base
+    <RequiredMember>
```

Comment:
Are we going to disallow usage of this attribute in VB?

Reply:
I wasn't planning on it.

Reply:
Added a note to test plan